cmd/ncproxy/main.go

```diff
@@ -0,0 +1,124 @@
+package main
+
+import (
+	"context"
+	"flag"
+	"net"
+	"os"
+	"os/signal"
+	"strings"
+	"syscall"
+
+	"github.com/Microsoft/go-winio"
+	"github.com/Microsoft/hcsshim/cmd/ncproxy/ncproxygrpc"
+	"github.com/Microsoft/hcsshim/internal/ncproxyttrpc"
+
+	"github.com/Microsoft/hcsshim/cmd/ncproxy/configagent"
+	"github.com/Microsoft/hcsshim/internal/computeagent"
+	"github.com/Microsoft/hcsshim/internal/log"
+	"github.com/sirupsen/logrus"
+)
+
+var (
+	configPath = flag.String("config", "", "Path to JSON configuration file.")
+	// Global mapping of network namespace ID to shim compute agent ttrpc service.
+	namespaceToShim = make(map[string]computeagent.ComputeAgentService)
+	// Mapping of network name to config agent clients
+	networkToConfigAgent = make(map[string]configagent.NetworkConfigAgentClient)
+)
+
+func main() {
+	flag.Parse()
+	ctx := context.Background()
+	config, err := loadConfig(*configPath)
+	if err != nil {
+		log.G(ctx).WithError(err).Error("failed getting configuration file")
+		os.Exit(1)
+	}
+
+	if config.GRPCAddr == "" {
+		log.G(ctx).Error("missing GRPC endpoint in config")
+		os.Exit(1)
+	}
+	if config.TTRPCAddr == "" {
+		log.G(ctx).Error("missing TTRPC endpoint in config")
+		os.Exit(1)
+	}
+
+	// Construct config agent client connections from config file information.
+	if err := configToClients(ctx, config); err != nil {
+		log.G(ctx).WithError(err).Error("failed to dial clients")
+		os.Exit(1)
+	}
+
+	log.G(ctx).Info("Starting ncproxy")
+
+	sigChan := make(chan os.Signal, 1)
+	serveErr := make(chan error, 1)
+	defer close(serveErr)
+	signal.Notify(sigChan, syscall.SIGINT)
+	defer signal.Stop(sigChan)
```

Comment:
do we also need to close sigChan?

Reply:
I don't even think the close of serveErr is necessary tbh. https://nanxiao.gitbooks.io/golang-101-hacks/content/posts/need-not-close-every-channel.html